### PR TITLE
Fix keys name for Transformer

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

This PR solves an issue due to #2489. Indeed, this PR modify the name of a key in decoder transformer `self.mutihead_attn` to `self.multihead_attn`. Doing so breaks the loading of state dict since it does not recognize the previous key. In order to solve this problem, I introduce a new function called `map_old_state_dict_weights` which is directly applied within `torch_recovery`, `average_checkpoints` and `_load_from_state_dict`. The first is when you are loading a checkpoint, the second when you are trying to avg multiple checkpoints (the issue here is that you need to make sure that before doing the avg, every ckpts has the same keys), and the latter when you are loading the state_dict directly from the object which can be the case in our codebase (i.e. bypassing checkpointer). 

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/utils/checkpoints.py [hunk 1]
```diff
@@ -94,10 +131,13 @@ def torch_recovery(obj, path, end_of_epoch):
     """
     del end_of_epoch  # Unused
     device = "cpu"
+
+    state_dict = torch.load(path, map_location=device)
+    state_dict = map_old_state_dict_weights(state_dict, KEYS_MAPPING)
```
Comment (asumagic) [inline speechbrain/utils/checkpoints.py:136]: Shouldn't we have a more generic `state_dict` fixup function that would call into `map_old_state_dict_weights`, even if the general function only calls that one for now?

File: speechbrain/utils/checkpoints.py [hunk 2]
```diff
@@ -75,6 +77,41 @@
...
+    """
+    Maps the keys in the old state dictionary according to the provided mapping,
+    under the given prefix.
+
+    Parameters
+    ----------
+    state_dict : dict
+        The old state dictionary to be mapped.
+    mapping : dict
+        A dictionary specifying the mapping between old and new keys.
```
Comment (asumagic) [inline speechbrain/utils/checkpoints.py:98]: It is unclear the extent of what this is, in particular, to my understanding:

- This can affect any part of the key, which could look like `somemodule.somesubmodule.somefield` rather than just `somefield`
- This will do partial matches on keys (e.g. `"a": "b"` will turn `_a` into `_b`)

The behavior should be documented to avoid accidentally including renames with too many side effects.

File: speechbrain/utils/checkpoints.py [hunk 3]
```diff
@@ -75,6 +77,41 @@
...
+
+    Returns
+    -------
+    dict
+        The modified state dictionary with mapped keys.
+    """
+    for checkpoint_name, attribute_name in mapping.items():
+        for full_checkpoint_name in list(state_dict.keys()):
+            if checkpoint_name in full_checkpoint_name:
+                full_attribute_name = full_checkpoint_name.replace(
```
Comment (asumagic) [inline speechbrain/utils/checkpoints.py:108]: Should we have a warning here?

File: speechbrain/utils/checkpoints.py (hunk 1 above)
Comment (Adel-Moumen) [reply speechbrain/utils/checkpoints.py:136]: I don't understand your question sorry haha

Comment (asumagic) [reply speechbrain/utils/checkpoints.py:136]: Having a function where we could add other `state_dict` fixes in the future.

File: speechbrain/utils/checkpoints.py (hunk 3 above)
Comment (Adel-Moumen) [reply speechbrain/utils/checkpoints.py:108]: Yep. Either a warning or a log in debug mode.

File: speechbrain/utils/checkpoints.py (hunk 1 above)
Comment (Adel-Moumen) [reply speechbrain/utils/checkpoints.py:136]: so you mean something like:
```python
def foo(state_dict):
	state_dict(state_dict)
```

so that someone can do : 
```python
def foo(state_dict):
	bar(state_dict)
	state_dict(state_dict)
```

easily?

**Comment by asumagic** [comment]
Seems like the PR current doesn't resolve issues with HF models loaded through `inference`, see https://github.com/speechbrain/speechbrain/issues/2549#issuecomment-2149528797

**Comment by asumagic** [comment]
> Seems like the PR current doesn't resolve issues with HF models loaded through `inference`, see [#2549 (comment)](https://github.com/speechbrain/speechbrain/issues/2549#issuecomment-2149528797)

To solve this, I made `torch_parameter_transfer` use the same logic, which I think is OK.

Other potentially problematic calls:

- `average_checkpoints()` in `checkpoints.py` defaults to `parameter_loader=torch.load`.

Other calls that do load model `state_dict`s(?) but _shouldn't be an issue_:

- In `_cycliclrloader`

Other calls that use `torch.load` for other things that _shouldn't be an issue_:

- In `InputNormalization` (arbitrary dict)
- In `schedulers.py` (arbitrary dict)
- In `huggingface_transformers/*` (non-SB `state_dict`s)
- In `k2_integration/*` (seems like graphs and other k2 stuff)
- In `inference/TTS.py` (arbitrary dict)
- In `aligner.py` (arbitrary dict)
- In `recipes/Voicebank/ASR/CTC/train.py` (not used and local, probably not a worry)

**Comment by Adel-Moumen** [comment]
sounds good to me.

**Review by asumagic — APPROVED** [review]
